File: t/PACSFramework/CS/StorageServerManager/UI/ServerSettings/ServerNetworking/ServerNetworkingPresenter.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Leadtools.Demos.StorageServer.DataTypes;
using Leadtools.Dicom.AddIn.Common;
using Leadtools.Medical.Winforms;

namespace Leadtools.Demos.StorageServer.UI
{
   public class ServerNetworkingPresenter
   {
      #region Public
         
         #region Methods
         
            public void RunView ( ServerNetworkingView view ) 
            {
               ServerState.Instance.ServerServiceChanged += new EventHandler  ( OnConfigureView ) ;
               ServerState.Instance.ServiceAdminChanged  += new EventHandler  ( OnConfigureView ) ;
               ServerState.Instance.IsRemoteServerChanged += new EventHandler ( OnConfigureView ) ;
               
               EventBroker.Instance.Subscribe <ApplyServerSettingsEventArgs>  ( OnUpdateServerSettings ) ;
               EventBroker.Instance.Subscribe <CancelServerSettingsEventArgs> ( OnCancelServerSettings ) ;

               View = view ;

               ConfigureView ( ) ;

               View.SettingsChanged += new EventHandler(View_SettingsChanged);
            }
            
            public void UpdateSettings ( ) 
            {
               if ( ServerState.Instance.ServerService != null ) 
               {
                  UpdateServerSettings ( ServerState.Instance.ServerService.Settings ) ;

                  ServerState.Instance.ServerService.Settings = ServerState.Instance.ServerService.Settings ;
               }
            }
         
         #endregion
         
         #region Properties
         
            public ServerNetworkingView View { get; private set; }
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
         #region Callbacks
            
         #endregion
         
      #endregion
      
      #region Protected
         
         #region Methods
            
         #endregion
         
         #region Properties
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
      #endregion
      
      #region Private
         
         #region Methods
         
            private void ConfigureView ( ) 
            {
               if ( ServerState.Instance.IsRemoteServer )
               {
                  View.Enabled = false ;
                  
                  return ;
               }
               
               if ( ServerState.Instance.ServerService != null ) 
               {
                  View.IsMaxPduLengthChecked = ServerState.Instance.ServerService.Settings.MaxPduLength != -1 ;
                  View.MaxPduLength          = ServerState.Instance.ServerService.Settings.MaxPduLength == -1 ? 0 : ServerState.Instance.ServerService.Settings.MaxPduLength ;
                  View.ReceiveBufferSize     = ServerState.Instance.ServerService.Settings.ReceiveBufferSize ;
                  View.SendBufferSize        = ServerState.Instance.ServerService.Settings.SendBufferSize ;
                  View.NoDelay               = ServerState.Instance.ServerService.Settings.NoDelay ;
               }
               else
               {
                  View.IsMaxPduLengthChecked = false ;
                  View.MaxPduLength          = 0;
                  View.ReceiveBufferSize     = 29696 ;
                  View.SendBufferSize        = 29696 ;
                  View.NoDelay               = false ;
               }

               __IsDirty = false ;
            }

            private void UpdateServerSettings ( ServerSettings serverSettings )
            {
               serverSettings.MaxPduLength = View.IsMaxPduLengthChecked ? View.MaxPduLength : -1 ;
               serverSettings.ReceiveBufferSize = View.ReceiveBufferSize ;
               serverSettings.SendBufferSize = View.SendBufferSize ;
               serverSettings.NoDelay = View.NoDelay ;
            }
            
         #endregion
         
         #region Properties
         
            private bool __IsDirty { get; set; }
            
         #endregion
         
         #region Events
         
            void OnConfigureView ( object sender, EventArgs e )
            {
               ConfigureView ( ) ;
            }

            public event EventHandler SettingsChanged;
            
            void View_SettingsChanged ( object sender, EventArgs e )
            {
               try
               {
                  if (SettingsChanged != null)
                  {
                     __IsDirty = true;
                     SettingsChanged(sender, e);
                  }
               }
               catch (Exception)
               {
                  System.Diagnostics.Debug.Assert(false);
               }
            }

            void OnUpdateServerSettings ( object sender, EventArgs e ) 
            {
               if ( __IsDirty ) 
               {
                  UpdateSettings ( ) ;
               }
            }

            void OnCancelServerSettings ( object sender, EventArgs e ) 
            {
               if ( __IsDirty ) 
               {
                  ConfigureView ( ) ;
               }
            }
         
         #endregion
         
         #region Data Members
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
      #endregion
      
      #region internal
         
         #region Methods
            
         #endregion
         
         #region Properties
            
         #endregion
         
         #region Data Types Definition
            
         #endregion
         
         #region Callbacks
            
         #endregion
         
      #endregion
   }
}

File: t/PACSFramework/VB/StorageServerManager/UI/ServerSettings/ServerNetworking/ServerNetworkingPresenter.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports Leadtools.Demos.StorageServer.DataTypes
Imports Leadtools.Dicom.AddIn.Common
Imports Leadtools.Medical.Winforms
Imports Leadtools.Demos.StorageServer

Namespace Leadtools.Demos.StorageServer.UI
   Public Class ServerNetworkingPresenter
#Region "Public"

#Region "Methods"

      Public Sub RunView(ByVal viewParam As ServerNetworkingView)
         AddHandler ServerState.Instance.ServerServiceChanged, AddressOf OnConfigureView
         AddHandler ServerState.Instance.ServiceAdminChanged, AddressOf OnConfigureView
         AddHandler ServerState.Instance.IsRemoteServerChanged, AddressOf OnConfigureView

         EventBroker.Instance.Subscribe(Of ApplyServerSettingsEventArgs)(AddressOf OnUpdateServerSettings)
         EventBroker.Instance.Subscribe(Of CancelServerSettingsEventArgs)(AddressOf OnCancelServerSettings)

         View = viewParam

         ConfigureView()

         AddHandler View.SettingsChanged, AddressOf View_SettingsChanged
      End Sub

      Public Sub UpdateSettings()
         If Not ServerState.Instance.ServerService Is Nothing Then
            UpdateServerSettings(ServerState.Instance.ServerService.Settings)

            ServerState.Instance.ServerService.Settings = ServerState.Instance.ServerService.Settings
         End If
      End Sub

#End Region

#Region "Properties"

      Private _view As ServerNetworkingView
      Public Property View() As ServerNetworkingView
         Get
            Return _view
         End Get
         Private Set(ByVal value As ServerNetworkingView)
            _view = value
         End Set
      End Property
#End Region

#Region "Data Types Definition"

#End Region

#Region "Callbacks"

#End Region

#End Region

#Region "Protected"

#Region "Methods"

#End Region

#Region "Properties"

#End Region

#Region "Data Types Definition"

#End Region

#End Region

#Region "Private"

#Region "Methods"

      Private Sub ConfigureView()
         If ServerState.Instance.IsRemoteServer Then
            View.Enabled = False

            Return
         End If

         If Not ServerState.Instance.ServerService Is Nothing Then
            View.IsMaxPduLengthChecked = ServerState.Instance.ServerService.Settings.MaxPduLength <> -1
            If ServerState.Instance.ServerService.Settings.MaxPduLength = -1 Then
               View.MaxPduLength = 0
            Else
               View.MaxPduLength = ServerState.Instance.ServerService.Settings.MaxPduLength
            End If
            View.ReceiveBufferSize = ServerState.Instance.ServerService.Settings.ReceiveBufferSize
            View.SendBufferSize = ServerState.Instance.ServerService.Settings.SendBufferSize
            View.NoDelay = ServerState.Instance.ServerService.Settings.NoDelay
         Else
            View.IsMaxPduLengthChecked = False
            View.MaxPduLength = 0
            View.ReceiveBufferSize = 29696
            View.SendBufferSize = 29696
            View.NoDelay = False
         End If

         __IsDirty = False
      End Sub

      Private Sub UpdateServerSettings(ByVal serverSettings As ServerSettings)
         If View.IsMaxPduLengthChecked Then
            serverSettings.MaxPduLength = View.MaxPduLength
         Else
            serverSettings.MaxPduLength = -1
         End If
         serverSettings.ReceiveBufferSize = View.ReceiveBufferSize
         serverSettings.SendBufferSize = View.SendBufferSize
         serverSettings.NoDelay = View.NoDelay
      End Sub

#End Region

#Region "Properties"

      Private _myIsDirty As Boolean
      Private Property __IsDirty() As Boolean
         Get
            Return _myIsDirty
         End Get
         Set(ByVal value As Boolean)
            _myIsDirty = value
         End Set
      End Property

#End Region

#Region "Events"

      Sub OnConfigureView(ByVal sender As Object, ByVal e As EventArgs)
         ConfigureView()
      End Sub

      Public Event SettingsChanged As EventHandler

      Sub View_SettingsChanged(ByVal sender As Object, ByVal e As EventArgs)
         Try
            __IsDirty = True
            RaiseEvent SettingsChanged(sender, e)
         Catch e1 As Exception
            System.Diagnostics.Debug.Assert(False)
         End Try
      End Sub

      Sub OnUpdateServerSettings(ByVal sender As Object, ByVal e As EventArgs)
         If __IsDirty Then
            UpdateSettings()
         End If
      End Sub

      Sub OnCancelServerSettings(ByVal sender As Object, ByVal e As EventArgs)
         If __IsDirty Then
            ConfigureView()
         End If
      End Sub

#End Region

#Region "Data Members"

#End Region

#Region "Data Types Definition"

#End Region

#End Region

#Region "internal"

#Region "Methods"

#End Region

#Region "Properties"

#End Region

#Region "Data Types Definition"

#End Region

#Region "Callbacks"

#End Region

#End Region
   End Class
End Namespace
